### PR TITLE
cortx-33340

### PR DESCRIPTION
**Pull request by ApurwaMohite**
Signed-off-by: ApurwaMohite <55824145+ApurwaMohite@users.noreply.github.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 1]
```diff
@@ -717,3 +718,169 @@ def test_reads_during_server_pod_restart(self):
...
+        responses_wr = dict()
+        while len(responses_wr) != 2:
+            responses_wr = output_wr.get(timeout=HA_CFG["common_params"]["60sec_delay"])
+        pass_logs = list(x[1] for x in responses_wr["pass_res"])
+        fail_logs = list(x[1] for x in responses_wr["fail_res"])
+        resp = self.ha_obj.check_s3bench_log(file_paths=pass_logs)
+        assert_utils.assert_false(len(resp[1]), f"WRITEs logs which contain failures: {resp[1]}")
+        resp = self.ha_obj.check_s3bench_log(file_paths=fail_logs, pass_logs=False)
+        LOGGER.debug("WRITEs Response for fail logs: %s", resp)
+        # TODO: Uncomment following once CORTX-28541 is fixed and re-test
```
Comment (rahul-hatwar) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:836]: Do we this this? This is server pod restart

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 2]
```diff
@@ -717,3 +718,169 @@ def test_reads_during_server_pod_restart(self):
         assert_utils.assert_true(resp[0], resp[1])
         LOGGER.info("Step 8: Successfully ran IOs after server pod restart")
         LOGGER.info("ENDED: Test to verify READs during data pod restart.")
```
Comment (sonalk0209) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:720]: not part of ur PR but this line need changes for server pod restart instead of data

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 3]
```diff
@@ -717,3 +718,169 @@ def test_reads_during_server_pod_restart(self):
...
+
+    @pytest.mark.ha
+    @pytest.mark.lc
+    @pytest.mark.tags("TEST-34261")
+    def test_server_pod_restart_kubectl_delete(self):
+        """
+        Verify IOs during pod restart (kubectl delete)
+        """
+        LOGGER.info("STARTED: Verify IOs after server pod restart (kubectl delete)")
```
Comment (sonalk0209) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:729]: Is it after or during? line 729 says after and 727 says during?

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 4]
```diff
@@ -717,3 +718,169 @@ def test_reads_during_server_pod_restart(self):
...
+        while len(responses_wr) != 2:
+            responses_wr = output_wr.get(timeout=HA_CFG["common_params"]["60sec_delay"])
+        pass_logs = list(x[1] for x in responses_wr["pass_res"])
+        fail_logs = list(x[1] for x in responses_wr["fail_res"])
+        resp = self.ha_obj.check_s3bench_log(file_paths=pass_logs)
+        assert_utils.assert_false(len(resp[1]), f"WRITEs logs which contain failures: {resp[1]}")
+        resp = self.ha_obj.check_s3bench_log(file_paths=fail_logs, pass_logs=False)
+        LOGGER.debug("WRITEs Response for fail logs: %s", resp)
+        # TODO: Uncomment following once CORTX-28541 is fixed and re-test
+        # assert_utils.assert_true(len(resp[1]) != len(fail_logs), "Some In-flight WRITEs are "
```
Comment (sonalk0209) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:837]: In this case, we may have to add condition for "may or may not" see failures as if there are any IOs during pod going down, we might get failures but while pod coming up we won't and we are clearing event after pod comes up and we check cluster status to know all services up.

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 5]
```diff
@@ -717,3 +718,169 @@ def test_reads_during_server_pod_restart(self):
...
+        fail_del_bkt = del_resp[1]  # Contains buckets which failed when event was clear
+        assert_utils.assert_false(len(fail_del_bkt), "Expected pass, buckets which failed in "
+                                                     f"DELETEs: {fail_del_bkt}.")
+        # TODO: Uncomment following once CORTX-28541 is fixed and re-test
+        # assert_utils.assert_true(len(event_del_bkt), "Expected FAIL when event was set")
+        LOGGER.info("Failed buckets while in-flight DELETEs operation : %s", event_del_bkt)
+        LOGGER.info("Step 5.3: Verified responses from DELETEs background process")
+        LOGGER.info("Step 5: Verified responses from background processes")
+
+        LOGGER.info("Step 6: Start IOs (create s3 acc, buckets and upload objects) after pod "
```
Comment (sonalk0209) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:876]: create IAM user

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 6]
```diff
@@ -717,3 +718,169 @@ def test_reads_during_server_pod_restart(self):
...
+                    "restart by kubectl delete")
+        users = self.mgnt_ops.create_account_users(nusers=1)
+        self.s3_clean.update(users)
+        self.test_prefix = 'test-34261'
+        resp = self.ha_obj.ha_s3_workload_operation(s3userinfo=list(users.values())[0],
+                                                    log_prefix=self.test_prefix,
+                                                    setup_s3bench=False)
+        assert_utils.assert_true(resp[0], resp[1])
+        LOGGER.info("Step 6: Successfully IOs completed after pod restart by kubectl delete")
+        LOGGER.info("COMPLETED: Verify IOs after server pod restart (kubectl delete)")
```
Comment (sonalk0209) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:886]: *during

File: tests/ha/single_pod_restart/test_server_pod_restart_api.py [hunk 7]
```diff
@@ -768,3 +771,159 @@ def test_copy_obj_after_server_pod_restart(self):
...
+        LOGGER.info("Step 2: Successfully performed chuck upload")
+
+        workload_info[1] = [self.bucket_name, self.object_name]
+
+        LOGGER.info("Calculating checksum of uploaded file %s", chunk_obj_path)
+        upld_chksm_hlt = self.ha_obj.cal_compare_checksum(file_list=[chunk_obj_path],
+                                                          compare=False)[0]
+
+        num_replica = self.num_replica - 1
+        LOGGER.info("Step 3: Shutdown server pod by making replicas=0 and verify cluster & "
```
Comment (sonalk0209) [inline tests/ha/single_pod_restart/test_server_pod_restart_api.py:823]: *with replica method

File: tests/ha/single_pod_restart/test_server_pod_restart_api.py [hunk 8]
```diff
@@ -768,3 +771,159 @@ def test_copy_obj_after_server_pod_restart(self):
...
+        resp = self.ha_obj.delete_kpod_with_shutdown_methods(
+            master_node_obj=self.node_master_list[0], health_obj=self.hlth_master_list[0],
+            pod_prefix=[const.SERVER_POD_NAME_PREFIX], delete_pod=[self.delete_pod],
+            num_replica=num_replica)
+        # Assert if empty dictionary
+        assert_utils.assert_true(resp[1], "Failed to shutdown/delete pod")
+        pod_name = list(resp[1].keys())[0]
+        if self.set_type == const.STATEFULSET:
+            self.set_name = resp[1][pod_name]['deployment_name']
+        elif self.set_type == const.REPLICASET:
```
Comment (sonalk0209) [inline tests/ha/single_pod_restart/test_server_pod_restart_api.py:834]: do we really need this if we are going to have sts only?

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 9]
```diff
@@ -717,3 +718,169 @@ def test_reads_during_server_pod_restart(self):
...
+
+        LOGGER.info("Step 1.2: Perform WRITEs with variable object sizes for parallel READs")
+        test_prefix_read = f'test-read-34261-{t_t}'
+        resp = self.ha_obj.ha_s3_workload_operation(s3userinfo=list(users.values())[0],
+                                                    log_prefix=test_prefix_read, skipread=True,
+                                                    skipcleanup=True)
+        assert_utils.assert_true(resp[0], resp[1])
+        LOGGER.info("Step 1.2: Performed WRITEs with variable sizes objects for parallel READs.")
+
+        LOGGER.info("Step 2.1: Start WRITEs with variable object sizes in background")
```
Comment (ankita-93) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:764]: Logger for step 2.
Starting background IOs ....

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 10]
```diff
@@ -717,3 +718,169 @@ def test_reads_during_server_pod_restart(self):
...
+        args = {'test_prefix': test_prefix_del, 'test_dir_path': self.test_dir_path,
+                'skipput': True, 'skipget': True, 'bkt_list': del_buckets,
+                'bkts_to_del': del_bucket, 'output': del_output}
+        thread_del = threading.Thread(target=self.ha_obj.put_get_delete,
+                                      args=(event, s3_test_obj,), kwargs=args)
+        thread_del.daemon = True  # Daemonize thread
+        thread_wri.start()
+        thread_rd.start()
+        thread_del.start()
+        LOGGER.info("Step 2.3: Started WRITEs, READs and DELETEs with variable object sizes "
```
Comment (ankita-93) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:799]: Or make this step 2.

File: tests/ha/single_pod_restart/test_server_pod_restart_api.py [hunk 11]
```diff
@@ -768,3 +771,159 @@ def test_copy_obj_after_server_pod_restart(self):
...
+        download_path = os.path.join(self.test_dir_path, "test_chunk_upload" + "_download")
+        chunk_obj_path = os.path.join(self.test_dir_path, self.object_name)
+        upload_op = Queue()
+        workload_info = dict()
+
+        LOGGER.info("Step 1: Perform setup steps for jclient")
+        jc_obj = JCloudClient()
+        resp = self.ha_obj.setup_jclient(jc_obj)
+        assert_utils.assert_true(resp, "Failed in setting up jclient")
+        LOGGER.info("Step 1: Successfully setup jcloud/jclient on runner")
```
Comment (ankita-93) [inline tests/ha/single_pod_restart/test_server_pod_restart_api.py:793]: jclient only right

File: tests/ha/single_pod_restart/test_server_pod_restart_api.py [hunk 12]
```diff
@@ -768,3 +771,159 @@ def test_copy_obj_after_server_pod_restart(self):
...
+
+        resp = self.ha_obj.create_bucket_chunk_upload(s3_data=self.s3_clean,
+                                                      bucket_name=self.bucket_name,
+                                                      file_size=file_size,
+                                                      chunk_obj_path=chunk_obj_path,
+                                                      background=False)
+        self.extra_files.append(chunk_obj_path)
+        assert_utils.assert_true(resp, "Failure observed in chunk upload in healthy cluster")
+        LOGGER.info("Step 2: Successfully performed chuck upload")
+
```
Comment (ankita-93) [inline tests/ha/single_pod_restart/test_server_pod_restart_api.py:815]: Reduce use of blank line here

File: tests/ha/single_pod_restart/test_server_pod_restart_api.py (hunk 8 above)
Comment (ApurwaMohite) [reply tests/ha/single_pod_restart/test_server_pod_restart_api.py:834]: Ok, we will assume that server and data will be of statefulset.

File: tests/ha/single_pod_restart/test_server_pod_restart.py (hunk 4 above)
Comment (ApurwaMohite) [reply tests/ha/single_pod_restart/test_server_pod_restart.py:837]: Can't we just print the result? Because adding condition for "may or may not" won't add any value.

File: tests/ha/single_pod_restart/test_server_pod_restart.py (hunk 2 above)
Comment (ApurwaMohite) [reply tests/ha/single_pod_restart/test_server_pod_restart.py:720]: @rahul-hatwar  took care of this in PR#1803

File: tests/ha/single_pod_restart/test_server_pod_restart.py (hunk 1 above)
Comment (ApurwaMohite) [reply tests/ha/single_pod_restart/test_server_pod_restart.py:840]: Yes, because this test covers pod shutdown and restart. We expect failures while pod is going down.

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 13]
```diff
@@ -717,3 +718,171 @@ def test_reads_during_server_pod_restart(self):
...
+        LOGGER.info("Background WRITEs, READs and DELETEs threads joined successfully.")
+        LOGGER.info("Step 5: Verify responses from background processes")
+        LOGGER.info("Step 5.1: Verify status for In-flight WRITEs")
+        responses_wr = dict()
+        while len(responses_wr) != 2:
+            responses_wr = output_wr.get(timeout=HA_CFG["common_params"]["60sec_delay"])
+        pass_logs = list(x[1] for x in responses_wr["pass_res"])
+        fail_logs = list(x[1] for x in responses_wr["fail_res"])
+        resp = self.ha_obj.check_s3bench_log(file_paths=pass_logs)
+        assert_utils.assert_false(len(resp[1]), f"WRITEs logs which contain failures: {resp[1]}")
```
Comment (sonalk0209) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:837]: maybe we can enhance this error message saying Writes not expected to fail before and after pod restarted, same for others

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 14]
```diff
@@ -717,3 +718,171 @@ def test_reads_during_server_pod_restart(self):
...
+        del_resp = tuple()
+        while len(del_resp) != 2:
+            del_resp = del_output.get(timeout=HA_CFG["common_params"]["60sec_delay"])
+        if not del_resp:
+            assert_utils.assert_true(False, "Background process failed to do DELETEs")
+        event_del_bkt = del_resp[0]  # Contains buckets when event was set
+        fail_del_bkt = del_resp[1]  # Contains buckets which failed when event was clear
+        assert_utils.assert_false(len(fail_del_bkt), "Expected pass, buckets which failed in "
+                                                     f"DELETEs: {fail_del_bkt}.")
+        # TODO: Uncomment following once CORTX-28541 is fixed and re-test
```
Comment (sonalk0209) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:872]: lets remove this as well, we can just print in case we have any deletes till we get some dry runs done

**Review by sonalk0209 — APPROVED** [review]
Minor comments, please check

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 15]
```diff
@@ -717,3 +718,171 @@ def test_reads_during_server_pod_restart(self):
...
+                                                    skipcleanup=True)
+        assert_utils.assert_true(resp[0], resp[1])
+        LOGGER.info("Step 1.2: Performed WRITEs with variable sizes objects for parallel READs.")
+
+        LOGGER.info("Step 2: Start WRITEs, READs and DELETEs with variable object sizes "
+                    "during server pod restart using kubectl delete")
+        LOGGER.info("Step 2.1: Start WRITEs with variable object sizes in background")
+        test_prefix_write = f'test-write-34261-{t_t}'
+        output_wr = Queue()
+        event_set_clr = [False]
```
Comment (rahul-hatwar) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:769]: is this require?

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 16]
```diff
@@ -717,3 +718,171 @@ def test_reads_during_server_pod_restart(self):
...
+        # assert_utils.assert_true(len(event_del_bkt), "Expected FAIL when event was set")
+        LOGGER.info("Failed buckets while in-flight DELETEs operation : %s", event_del_bkt)
+        LOGGER.info("Step 5.3: Verified responses from DELETEs background process")
+        LOGGER.info("Step 5: Verified responses from background processes")
+
+        LOGGER.info("Step 6: Start IOs (create IAM user, buckets and upload objects) after pod "
+                    "restart by kubectl delete")
+        users = self.mgnt_ops.create_account_users(nusers=1)
+        self.s3_clean.update(users)
+        self.test_prefix = 'test-34261'
```
Comment (rahul-hatwar) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:882]: can we make test-34261-1 - for consistency

File: tests/ha/single_pod_restart/test_server_pod_restart.py (hunk 15 above)
Comment (ApurwaMohite) [reply tests/ha/single_pod_restart/test_server_pod_restart.py:769]: Yes, as this is IOs during restart.

File: tests/ha/single_pod_restart/test_server_pod_restart.py [hunk 17]
```diff
@@ -717,3 +718,171 @@ def test_reads_during_server_pod_restart(self):
...
+        fail_logs = list(x[1] for x in responses_rd["fail_res"])
+        resp = self.ha_obj.check_s3bench_log(file_paths=pass_logs)
+        assert_utils.assert_false(len(resp[1]),
+                                  f"READs/VerifyDI logs which contain failures: {resp[1]}")
+        resp = self.ha_obj.check_s3bench_log(file_paths=fail_logs, pass_logs=False)
+        LOGGER.info("READs Response for fail logs: %s", resp[1])
+        assert_utils.assert_true(len(resp[1]) <= len(fail_logs), "Some In-flight READs are "
+                                                                 "expected to fail. No failure is"
+                                                                 f"observed in {resp[1]}")
+
```
Comment (rahul-hatwar) [inline tests/ha/single_pod_restart/test_server_pod_restart.py:859]: can we remove this extra lines between steps? Please have a look at other places too

File: tests/ha/single_pod_restart/test_server_pod_restart_api.py [hunk 18]
```diff
@@ -768,3 +771,155 @@ def test_copy_obj_after_server_pod_restart(self):
...
+        args = {'s3_data': self.s3_clean, 'bucket_name': bucket_name,
+                'file_size': file_size, 'chunk_obj_path': chunk_obj_path, 'output': upload_op}
+
+        thread = threading.Thread(target=self.ha_obj.create_bucket_chunk_upload, kwargs=args)
+        thread.daemon = True  # Daemonize thread
+        thread.start()
+        time.sleep(HA_CFG["common_params"]["30sec_delay"])
+        LOGGER.info("Step 5: Successfully started chuck upload in background")
+
+        LOGGER.info("Step 6: Start server pod again by statefulset")
```
Comment (rahul-hatwar) [inline tests/ha/single_pod_restart/test_server_pod_restart_api.py:868]: please change this LOGGER

File: tests/ha/single_pod_restart/test_server_pod_restart.py (hunk 16 above)
Comment (ApurwaMohite) [reply tests/ha/single_pod_restart/test_server_pod_restart.py:882]: No needed, because this prefix is not used in test above.

File: tests/ha/single_pod_restart/test_server_pod_restart.py (hunk 13 above)
Comment (ApurwaMohite) [reply tests/ha/single_pod_restart/test_server_pod_restart.py:837]: Yes, message can be enhanced. We will have to do this at all the places. Because almost at all places we are using same error message.

File: tests/ha/single_pod_restart/test_server_pod_restart.py (hunk 14 above)
Comment (ApurwaMohite) [reply tests/ha/single_pod_restart/test_server_pod_restart.py:872]: I think this TODO is applicable for deletes. Because it may happen that deletes are not failing during pod delete.
Actually it may or may not be applicable for reads and writes as well, depending on the edge condition. What do you suggest?

Comment (sonalk0209) [reply tests/ha/single_pod_restart/test_server_pod_restart.py:876]: As we had discussed lets add may or may not condition for all and lets print those logs and try to take multiple runs to see how much time it takes and what all edge conditions we meet.

**Review by ankita-93 — APPROVED** [review]

File: tests/ha/single_pod_restart/test_server_pod_restart_api.py [hunk 19]
```diff
@@ -768,3 +771,155 @@ def test_copy_obj_after_server_pod_restart(self):
...
+        object_name = f"chunk-upload-obj-{t_t}"
+        chunk_obj_path = os.path.join(self.test_dir_path, object_name)
+        LOGGER.info("Step 5: Start chunk upload in background")
+        args = {'s3_data': self.s3_clean, 'bucket_name': bucket_name,
+                'file_size': file_size, 'chunk_obj_path': chunk_obj_path, 'output': upload_op}
+
+        thread = threading.Thread(target=self.ha_obj.create_bucket_chunk_upload, kwargs=args)
+        thread.daemon = True  # Daemonize thread
+        thread.start()
+        time.sleep(HA_CFG["common_params"]["30sec_delay"])
```
Comment (rahul-hatwar) [inline tests/ha/single_pod_restart/test_server_pod_restart_api.py:865]: Add logger for wait

**Review by rahul-hatwar — APPROVED** [review]
Minor comment, LGTM

**Comment by ApurwaMohite** [comment]
Collection log:
[PR1798.log](https://github.com/Seagate/cortx-test/files/9298252/PR1798.log)

**Review by kachhwahadivya — APPROVED** [review]
Approving for merge
